examples/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,454 @@
+
+The dataset includes 48,842 instances with 14 input features, 5 of which numerical,
+and the other 9 are categorical.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+from tensorflow import keras
+from tensorflow.keras import layers
+import math
+
+"""
+## Prepare the data
+"""
+
+CSV_HEADER = [
+    "age",
+    "workclass",
+    "fnlwgt",
+    "education",
+    "education_num",
+    "marital_status",
+    "occupation",
+    "relationship",
+    "race",
+    "gender",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "native_country",
+    "income_bracket",
+]
+
+train_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+)
+train_data = pd.read_csv(train_data_url, header=None, names=CSV_HEADER)
+
+test_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test"
+)
+test_data = pd.read_csv(test_data_url, header=None, names=CSV_HEADER)
+
+print(f"Train dataset shape: {train_data.shape}")
+print(f"Test dataset shape: {test_data.shape}")
+
+train_data.head().T
+
+"""
+We remove the first record as it is not a valid data example, and we remove a tailing
+'dot' in the class labels.
+"""
+
+test_data = test_data[1:]
+test_data.income_bracket = test_data.income_bracket.apply(
+    lambda value: value.replace(".", "")
+)
+
+"""
+Now we store the train and test data splits locally to CSV files.
+"""
+
+train_data_file = "train_data.csv"
+test_data_file = "test_data.csv"
+
+train_data.to_csv(train_data_file, index=False, header=False)
+test_data.to_csv(test_data_file, index=False, header=False)
+
+"""
+## Define dataset metadata
+
+Here, we define the metadata of the dataset that will be useful for reading and parsing
+the data into input features, and encoding the input features with respect to their types.
+"""
+
+NUMERIC_FEATURE_NAMES = [
+    "age",
+    "education_num",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+]
+
+CATEGORICAL_FEATURES_WITH_VOCABULARY = {
+    "workclass": list(train_data["workclass"].unique()),
+    "education": list(train_data["education"].unique()),
+    "marital_status": list(train_data["marital_status"].unique()),
+    "occupation": list(train_data["occupation"].unique()),
+    "relationship": list(train_data["relationship"].unique()),
+    "race": list(train_data["race"].unique()),
+    "gender": list(train_data["gender"].unique()),
+    "native_country": list(train_data["native_country"].unique()),
+}
+
+IGNORE_COLUMN_NAMES = ["fnlwgt"]
+
+CATEGORICAL_FEATURE_NAMES = list(CATEGORICAL_FEATURES_WITH_VOCABULARY.keys())
+
+FEATURE_NAMES = NUMERIC_FEATURE_NAMES + CATEGORICAL_FEATURE_NAMES
+
+COLUMN_DEFAULTS = [
+    [0.0] if feature_name in NUMERIC_FEATURE_NAMES + IGNORE_COLUMN_NAMES else ["NA"]
+    for feature_name in CSV_HEADER
+]
+
+TARGET_FEATURE_NAME = "income_bracket"
+
+TARGET_LABELS = [" <=50K", " >50K"]
```

Comment:
Why the leading space?

Reply:
@ksalama Has this been clarified? Just checking

Reply:
This is how the data is in the original dataset.